### DISCUSSION TITLE
how to get column info of a __abstract__ = True orm class

### DISCUSSION THREAD

**Discussion by flyly0755** [sqlalchemy/sqlalchemy#11392] [Usage Questions] [answered]
With code showed as below, a orm clsss with __abstract__ = True attr, try with 2 methods to get column name info, but both failed.
```python
from sqlalchemy.types import Integer, String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, inspect

PgBase = declarative_base()


class TestTb1(PgBase):
    id = Column(Integer, autoincrement=True, primary_key=True)
    string_f1 = Column(String)
    int_f1 = Column(Integer)

    __abstract__ = True
    __tablename__ = 'TestTb1'


inst = inspect(TestTb1)
attr_names = [c_attr.key for c_attr in inst.mapper.column_attrs]   # sqlalchemy.exc.NoInspectionAvailable: No inspection system is available for object of type <class 'sqlalchemy.orm.decl_api.DeclarativeMeta'>

attr_names2 = TestTb1.__table__.columns.keys()  # AttributeError: type object 'TestTb1' has no attribute '__table__'
```
Is there any method to get column info(column name and column data type)?

Version:
SQLAlchemy==1.4.49
SQLAlchemy-Utils==0.41.1
psycopg2==2.9.6

**Comment by CaselIT** [comment, marked as answer]
Hi,

I think sqlalchemy will completely ignore the abstract classes. But you can probably use the standard python `vars(TestTb1)` to get a dict of attributes of the class

**Comment by zzzeek** [comment]
hi -

no, a class that's marked with `__abstract__` has no `Table` or `Mapper` present.

bigger issue, there should be no *need* to get this information from an `__abstract__` class.  such a class is only to be used at initial mapping time, not during any kind of introspection operation.
